docs/rfcs/2025-08-11-channels-forwarding.md

```diff
@@ -0,0 +1,143 @@
+- Members will deduplicate messages (already implemented).
+- Members should highlight differences in deduplicated messages. This would solve trust issue.
+  - Question: If messages are to be signed by owners, what differences can chat relays introduce? Does this point in initial doc imply owner sending different versions of message to chat relays? Losses in delivery?
+  - TBC prototype UI, backend.
+
+### Persisted forwarding instructions
+
+- For batch of received messages a forwarding instruction has to be persisted.
+- Special worker would pick up forwarding instructions for sending.
+- Forwarding instruction can be split into smaller batches by members to make smaller group sends one at a time.
+
+How to present a forwarding instruction in storage?
+
+Currently we build ad-hoc forwarding instruction for each message based on its scope (GroupForwardScope). Then for each scope message batches are sent separately to different member lists.
+
+Question: should we persist all forwarding instructions of or only for main (GFSMain - for regular messages) scope? For main and all (GFSAll), and don't persist for support scope (GFSMemberSupport)?
+
+We roughly need following data to be persisted for a forwarding instruction:
+- group id,
+- forwarding scope - TBC as above,
+- sending member - to include memberId in XGrpMsgForward,
+- broker timestamp of received messages batch to include in XGrpMsgForward,
+- list of messages (events).
+
+Restoring list of messages for a forwarding instruction:
+- Currently for each message full MsgBody is saved, which is just a ByteString. Also if received messages were batched, full batch body is saved for each message record. For example, here is message body of batched deletion of 2 chat items, which is saved on 2 message records:
+  ```
+  msg_body = [{"v":"1-16","msgId":"ZUd3bzVDTXFHWmc3Z29YSQ==","event":"x.msg.del","params":{"msgId":"QVVuN2RkaXg0aVJJdTZXSA=="}},{"v":"1-16","msgId":"amM5dzV2RkRjNmNYSDRLQQ==","event":"x.msg.del","params":{"msgId":"NzFlV3pNRStpQTRrQjRYUg=="}}]
+  ```
+  On the other hand for forwarding operation we require ChatMessage 'Json (for XGrpMsgForward).
+- For a forwarding instruction we can persist list of messages ids from messages table (message_id), and save ChatMessage 'Json on records to restore.
+- Alternatively, we can read full msg_body (it can even be duplicated on forwarding instruction), and restore chat messages (ChatMessage 'Json) via parseChatMessages. However, this conflicts with possibility that a single batch body will have messages of different scopes, which current processing logic allows, although there's no legitimate case as of now. So, persisting ChatMessage' Json on each record seems preferable. Also it avoids repeated parsing.
+- As a side note, it seems we can stop saving msg_body for received messages, as it's not used for any purpose. This field is only used for retrieving and sending pending group messages.
```

Comment:
I don't quite follow this argument, it's confusing.

Reply:
It discusses mechanism on how to save and restore messages for forwarding. Currently we save full batch body on each received message record (batch can create many received message records with the same batch body), restoring from which would require repeated parsing and does not account for possibility of messages with different scopes in a single batch.